File: 장세훈/LeetCode/medium/SwapNodesInPairs.java
class SwapNodesInPairs {
    /*
        24. Swap Nodes in Pairs
        https://leetcode.com/problems/swap-nodes-in-pairs/description/

        재귀로 풀이
        링크드 리스트의 노드들을 2개씩 짝으로 순서를 바꾼다.
        재귀를 이용해 노드의 다음 노드를 스왑한 노드로 바꿔준다.
     */


    public ListNode swapPairs(ListNode head) {
        return (head == null || head.next == null) ? head : swap(head);
    }

    public ListNode swap(ListNode node) {
        if (node == null) return null;

        ListNode first = node;
        ListNode second = node.next;
        if (second == null) return first;
        ListNode next = second.next == null ? null : second.next;

        second.next = first;
        first.next =  swap(next);

        return second;
    }
}

File: 정수빈/LeetCode/medium/swap-nodes-in-pairs.py
# https://leetcode.com/problems/swap-nodes-in-pairs/
# 걸린시간: 20분

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        node = head
        node_list = []
        while node:
            node_list.append(node)
            node = node.next
        
        for i in range(len(node_list)):
            if i % 2 == 0 and i+1 < len(node_list):
                tmp = node_list[i]
                node_list[i] = node_list[i+1]
                node_list[i+1] = tmp

        for i in range(len(node_list)-1):
            node_list[i].next = node_list[i+1]
            node_list[i+1].next = None

        return node_list[0] if node_list else None